odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2385 @@
+        )
+
+        Broadcasting and array-like input is supported, too:
+
+        >>> x.ufuncs.add(3)
+        rn((2, 3)).element(
+            [[4.0, 1.0, 6.0],
+             [7.0, -2.0, 9.0]]
+        )
+        >>> x.ufuncs.subtract([[0, 0, 1],
+        ...                   [1, 0, 0]])
+        rn((2, 3)).element(
+            [[1.0, -2.0, 2.0],
+             [3.0, -5.0, 6.0]]
+        )
+
+        There is also support for various reductions (sum, prod, min,
+        max):
+
+        >>> x.ufuncs.sum()
+        7.0
+
+        The ``out`` parameter can be used for result storage:
+
+        >>> y = space.one()
+        >>> out = space.element()
+        >>> result = x.ufuncs.add(y, out=out)
+        >>> result
+        rn((2, 3)).element(
+            [[2.0, -1.0, 4.0],
+             [5.0, -4.0, 7.0]]
+        )
+        >>> result is out
+        True
+        """
+        return NumpyTensorSetUfuncs(self)
+
+
+def _blas_is_applicable(*args):
+    """Whether BLAS routines can be applied or not.
+
+    BLAS routines are available for single and double precision
+    float or complex data only. If the arrays are non-contiguous,
+    BLAS methods are usually slower, and array-writing routines do
+    not work at all. Hence, only contiguous arrays are allowed.
+
+    Parameters
+    ----------
+    x1,...,xN : `NumpyGeneralizedTensor`
+        The tensors to be tested for BLAS conformity.
+    """
+    if any(x.dtype != args[0].dtype for x in args[1:]):
+        return False
+    if any(x.dtype not in _BLAS_DTYPES for x in args):
+        return False
+    if not (all(x.flags.f_contiguous for x in args) or
+            all(x.flags.c_contiguous for x in args)):
+        return False
+    return True
+
+
+def _lincomb(a, x1, b, x2, out, dtype):
+    """Raw linear combination depending on data type."""
+    # Shortcut for small problems
+    # if x1.size < 100:  # small array optimization
+    #     out.data[:] = a * x1.data + b * x2.data
+    #     return
+
+    # Use blas for larger problems
+    def fallback_axpy(x1, x2, n, a):
+        """Fallback axpy implementation avoiding copy."""
+        if a != 0:
+            x2 /= a
+            x2 += x1
+            x2 *= a
+        return x2
+
+    def fallback_scal(a, x, n):
+        """Fallback scal implementation."""
+        x *= a
+        return x
+
+    def fallback_copy(x1, x2, n):
+        """Fallback copy implementation."""
+        x2[...] = x1[...]
+        return x2
+
+    # Need flat data for BLAS, otherwise in-place does not work
+    x1_arr = x1.data.ravel(order=x1.space.view_order)
+    x2_arr = x2.data.ravel(order=x1.space.view_order)
+    out_arr = out.data.ravel(order=x1.space.view_order)
+
+    if _blas_is_applicable(x1.data, x2.data, out.data):
+        axpy, scal, copy = linalg.blas.get_blas_funcs(
+            ['axpy', 'scal', 'copy'], arrays=(x1.data, x2.data, out.data))
+    else:
+        axpy, scal, copy = (fallback_axpy, fallback_scal, fallback_copy)
```

Comment:
Actually, this should be baked into the space, or into the individual space elements (since it can vary for `'K'` spaces). All these checks do take quite a bit of time.

For `_lincomb` we also know everything about the `dtype` in advance, there should not be a need to check every single time.

Need to look at other use cases for the helper, though.